gdal/frmts/tiledb/tiledbdataset.cpp

```diff
@@ -0,0 +1,724 @@
+            break;
+        }
+        case GDT_CFloat64:
+        {
+            schema.add_attribute( tiledb::Attribute::create<double[2]>( *m_ctx, TILEDB_VALUES ) );
+            nBitsPerSample = 64;
+            nSampleFormat = SAMPLEFORMAT_IEEEFP;
+            break;
+        }
+        default:
+            return CE_Failure;
+    }
+    return CE_None;
+}
+
+/************************************************************************/
+/*                              Create()                                */
+/************************************************************************/
+
+GDALDataset *
+TileDBDataset::Create( const char * pszFilename, int nXSize, int nYSize, int nBands,
+        GDALDataType eType, char ** papszParmList )
+
+{
+    TileDBDataset *poDS = new TileDBDataset();
+    poDS->nRasterXSize = nXSize;
+    poDS->nRasterYSize = nYSize;
+    poDS->nBands = nBands;
+    poDS->eAccess = GA_Update;
+    poDS->eDataType = eType;
+
+    const char* pszConfig = CSLFetchNameValue( papszParmList, "CONFIG" );
+    if( pszConfig != nullptr )
+    {
+        tiledb::Config cfg( pszConfig );
+        poDS->m_ctx.reset( new tiledb::Context( cfg ) );
+    }
+    else
+    {
+        poDS->m_ctx.reset( new tiledb::Context() );
+    }
+
+    const char* pszCompression = CSLFetchNameValue( papszParmList, "COMPRESSION" );  
+    const char* pszCompressionLevel = CSLFetchNameValue( papszParmList, "COMPRESSION_LEVEL" );
+ 
+    const char* pszBlockXSize = CSLFetchNameValue( papszParmList, "BLOCKXSIZE" );
+    poDS->nBlockXSize = ( pszBlockXSize ) ? atoi( pszBlockXSize ) : 256;
+    const char* pszBlockYSize = CSLFetchNameValue( papszParmList, "BLOCKYSIZE" );
+    poDS->nBlockYSize = ( pszBlockYSize ) ? atoi( pszBlockYSize ) : 256;
+    poDS->bStats = CSLFetchBoolean( papszParmList, "STATS", FALSE );
+
+    for( int i = 0; i < poDS->nBands; i++ )
+        poDS->SetBand( i+1, new TileDBRasterBand( poDS, i+1 ) );
+
+    // set dimensions and attribute type for schema
+    tiledb::ArraySchema schema( *poDS->m_ctx, TILEDB_DENSE );
+    schema.set_tile_order( TILEDB_ROW_MAJOR );
+    schema.set_cell_order( TILEDB_ROW_MAJOR );
+
+    poDS->m_filterList.reset(new tiledb::FilterList(*poDS->m_ctx));
+
+    if (pszCompression != nullptr)
+    {
+        int nLevel = ( pszCompressionLevel ) ? atoi( pszCompressionLevel ) : -1;
+        poDS->AddFilter(pszCompression, nLevel);
+        schema.set_coords_filter_list(*poDS->m_filterList);
+    }
+
+    tiledb::Domain domain( *poDS->m_ctx );
+
+    // Note the dimension bounds are inclusive and are expanded to the match the block size
+    // TODO creating dims in tileb with int type fails with Static type (UINT64) does not match expected type (INT32)
+    size_t w = DIV_ROUND_UP(nXSize, poDS->nBlockXSize) * poDS->nBlockXSize;
+    size_t h = DIV_ROUND_UP(nYSize, poDS->nBlockYSize) * poDS->nBlockYSize;
+    
+    auto d1 = tiledb::Dimension::create<size_t>( *poDS->m_ctx, "X", {0, w}, poDS->nBlockXSize );
```

Comment:
If x is the first dimension (rows) and you read into gdal buffers in row-major order, won't that transpose the image (unless gdal buffers are expected to be x-major)?

Reply:
GDAL is row major, in GDAL a row corresponds to y and x to columns. From a quick test with `gdal_translate` going to TileDB and back I am not seeing a transpose.